homeassistant/components/zwave_js/event.py

```diff
@@ -0,0 +1,109 @@
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import ATTR_VALUE, DATA_CLIENT, DOMAIN
+from .discovery import ZwaveDiscoveryInfo
+from .entity import ZWaveBaseEntity
+
+PARALLEL_UPDATES = 0
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Z-Wave Event entity from Config Entry."""
+    client: ZwaveClient = hass.data[DOMAIN][config_entry.entry_id][DATA_CLIENT]
+
+    @callback
+    def async_add_event(info: ZwaveDiscoveryInfo) -> None:
+        """Add Z-Wave event entity."""
+        driver = client.driver
+        assert driver is not None  # Driver is ready before platforms are loaded.
+        entities: list[ZWaveBaseEntity] = []
+        if info.platform_hint == "stateless":
+            entities.append(ZwaveEventEntity(config_entry, driver, info))
+        else:
+            raise ValueError(
+                f"Discovered value with hint '{info.platform_hint} not handled"
+            )
+        async_add_entities(entities)
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(
+            hass,
+            f"{DOMAIN}_{config_entry.entry_id}_add_{EVENT_DOMAIN}",
+            async_add_event,
+        )
+    )
+
+
+def _cc_and_label(value: Value) -> str:
+    """Return a string with the command class and label."""
+    return f"{value.command_class_name}: {value.metadata.label}"
+
+
+class ZwaveEventEntity(ZWaveBaseEntity, EventEntity):
+    """Representation of a Z-Wave event entity."""
+
+    def __init__(
+        self, config_entry: ConfigEntry, driver: Driver, info: ZwaveDiscoveryInfo
+    ) -> None:
+        """Initialize a ZwaveEventEntity entity."""
+        super().__init__(config_entry, driver, info)
+
+        if info.primary_value.metadata.states:
+            self._attr_event_types = sorted(info.primary_value.metadata.states.values())
+        else:
+            self._attr_event_types = [_cc_and_label(info.primary_value)]
+        # Entity class attributes
+        self._attr_name = self.generate_name(include_value_name=True)
+
+    @callback
+    def _async_handle_event(self, value_notification: ValueNotification) -> None:
+        """Handle a value notification event."""
+        primary_value = self.info.primary_value
+        # If the notification doesn't match the value we are tracking, we can return
+        if (
+            value_notification.command_class != primary_value.command_class
+            or value_notification.endpoint != primary_value.endpoint
+            or value_notification.property_ != primary_value.property_
+            or value_notification.property_key != primary_value.property_key
+        ):
+            return
+
+        if primary_value.metadata.states:
+            try:
+                event_name = primary_value.metadata.states[
```

Comment:
Can the metadata states change after entity setup? It looks like we could create a map between the notification value and the event name already during entity setup. Then we wouldn't need all the logic here.

Reply:
I simplified in two ways:
1. Store a map we can access with the value already converted to an int
2. Assume that we will have all states and fallback to the event type for values without states when we can't look it up in the map